rng/src/main.rs

```diff
@@ -0,0 +1,170 @@
+
+// Chosen to replicate the max period found in QEMU's vhost-user-rng
+// and virtio-rng implementations.
+const VHU_RNG_MAX_PERIOD_MS: u128 = 65536;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, PartialEq, ThisError)]
+/// Errors related to vhost-device-rng daemon.
+pub enum Error {
+    #[error("RNG source file doesn't exists or can't be accessed")]
+    AccessRngSourceFile,
+    #[error("Period is too big")]
+    InvalidPeriodInput,
+    #[error("Socket path can't be found")]
+    InvalidSocketPath,
+    #[error("Wrong socket count")]
+    InvalidSocketCount,
+    #[error("Threads can't be joined")]
+    FailedJoiningThreads,
+}
+
+#[derive(Clone, Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct RngArgs {
+    /// Time needed (in ms) to transfer max-bytes amount of byte.
+    #[clap(short, long, default_value_t = VHU_RNG_MAX_PERIOD_MS)]
+    period: u128,
+
+    /// Maximum amount of byte that can be transferred in a period.
+    #[clap(short, long, default_value_t = usize::MAX)]
+    max_bytes: usize,
+
+    /// Number of guests (sockets) to connect to.
+    #[clap(short = 'c', long, default_value_t = 1)]
+    socket_count: u32,
+
+    /// Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
+    #[clap(short, long)]
+    socket_path: Option<String>,
+
+    /// Where to get the RNG data from. Defaults to /dev/urandom.
+    #[clap(short = 'f', long, default_value = "/dev/urandom")]
+    rng_source: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct VuRngConfig {
+    pub period_ms: u128,
+    pub max_bytes: usize,
+    pub count: u32,
+    pub socket_path: String,
+    pub rng_source: String,
+}
+
+impl TryFrom<RngArgs> for VuRngConfig {
+    type Error = Error;
+
+    fn try_from(args: RngArgs) -> Result<Self> {
+        if args.period > VHU_RNG_MAX_PERIOD_MS {
+            return Err(Error::InvalidPeriodInput);
+        }
+
+        if args.socket_count == 0 {
+            return Err(Error::InvalidSocketCount);
+        }
+
+        let socket_path = match args.socket_path {
+            Some(path) => path,
+            None => return Err(Error::InvalidSocketPath),
+        };
+
+        // Divide available bandwidth by the number of threads in order
+        // to avoid overwhelming the HW.
+        let max_bytes = args.max_bytes / args.socket_count as usize;
+
+        Ok(VuRngConfig {
+            period_ms: args.period,
+            max_bytes,
+            count: args.socket_count,
+            socket_path,
+            rng_source: args.rng_source,
+        })
+    }
+}
+
+pub fn start_backend(config: VuRngConfig) -> Result<()> {
+    let mut handles = Vec::new();
+    let random_file = Arc::new(Mutex::new(
+        File::open(&config.rng_source).map_err(|_| Error::AccessRngSourceFile)?,
+    ));
+
+    for i in 0..config.count {
+        let socket_path = config.socket_path.clone();
+        let socket = format!("{}{}", socket_path, i);
```

Comment:
You can also do this I think, instead of above two lines:

        let socket = format!("{}{}", config.socket_path.to_owned(), i);